In [ ]:
import heapq
import numpy as np

# --- 1. Define Maze (Figure 8) ---
# 10x10 Grid. 0 = Open, 1 = Obstacle
maze_fig8 = [
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0], # Row 0
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0], # Row 1
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0], # Row 2
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0], # Row 3
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0], # Row 4
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0], # Row 5
    [0, 0, 0, 0, 0, 0, 0, 0, 1, 0], # Row 6 (Obstacle at 6,8)
    [0, 0, 0, 0, 0, 0, 0, 1, 1, 0], # Row 7 (Obstacles at 7,7; 7,8)
    [0, 0, 0, 0, 0, 0, 1, 1, 1, 0], # Row 8 (Obstacles at 8,6; 8,7; 8,8)
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]  # Row 9
]

# --- 2. Helper Class & Functions ---

class Node:
    def __init__(self, position, parent=None):
        self.position = position
        self.parent = parent
        self.g = 0  # Cost from start
        self.h = 0  # Heuristic to goal
        self.f = 0  # Total cost

    # Comparison for Priority Queue (based on f value)
    def __lt__(self, other):
        return self.f < other.f

def manhattan_distance(current, goal):
    return abs(current[0] - goal[0]) + abs(current[1] - goal[1])

def get_neighbors(position, maze):
    """Returns valid neighbors (Up, Down, Left, Right)."""
    r, c = position
    rows = len(maze)
    cols = len(maze[0])
    moves = [(-1, 0), (1, 0), (0, -1), (0, 1)] # U, D, L, R

    neighbors = []
    for dr, dc in moves:
        nr, nc = r + dr, c + dc
        if 0 <= nr < rows and 0 <= nc < cols and maze[nr][nc] == 0:
            neighbors.append((nr, nc))
    return neighbors

def informed_search(maze, start, goal, strategy):
    # Create start node
    start_node = Node(start, None)
    start_node.g = 0
    start_node.h = manhattan_distance(start, goal)

    # Priority Queue
    open_list = []
    heapq.heappush(open_list, start_node)

    # Track visited nodes to prevent cycles
    visited = set()
    nodes_expanded = 0

    while open_list:
        # Get node with lowest 'f'
        current_node = heapq.heappop(open_list)
        nodes_expanded += 1

        # Check Goal
        if current_node.position == goal:
            path = []
            curr = current_node
            while curr:
                path.append(curr.position)
                curr = curr.parent
            return path[::-1], nodes_expanded

        # Skip if visited
        if current_node.position in visited:
            continue
        visited.add(current_node.position)

        # Explore Neighbors
        for neighbor_pos in get_neighbors(current_node.position, maze):
            if neighbor_pos in visited:
                continue

            new_node = Node(neighbor_pos, current_node)

            # Calculate Costs
            new_node.g = current_node.g + 1
            new_node.h = manhattan_distance(neighbor_pos, goal)

            # Strategy Logic
            if strategy == 'AStar':
                new_node.f = new_node.g + new_node.h
            elif strategy == 'GBF': # Greedy Best First
                new_node.f = new_node.h
            elif strategy == 'UCS': # Uniform Cost Search
                new_node.f = new_node.g

            heapq.heappush(open_list, new_node)

    return None, nodes_expanded

# --- 3. Execute Searches (9a) ---
start_pos = (0, 0)
goal_pos = (9, 9)

print(f"--- 9a. Search Results for Goal {goal_pos} ---")

# Run A*
path_astar, exp_astar = informed_search(maze_fig8, start_pos, goal_pos, 'AStar')
print(f"\n[A* Search]")
print(f"Path Length: {len(path_astar)-1} steps")
print(f"Nodes Expanded: {exp_astar}")
print(f"Path: {path_astar}")

# Run Greedy Best First (GBF)
path_gbf, exp_gbf = informed_search(maze_fig8, start_pos, goal_pos, 'GBF')
print(f"\n[Greedy Best First Search (GBF)]")
print(f"Path Length: {len(path_gbf)-1} steps")
print(f"Nodes Expanded: {exp_gbf}")
print(f"Path: {path_gbf}")

# Run Uniform Cost Search (UCS)
path_ucs, exp_ucs = informed_search(maze_fig8, start_pos, goal_pos, 'UCS')
print(f"\n[Uniform Cost Search (UCS)]")
print(f"Path Length: {len(path_ucs)-1} steps")
print(f"Nodes Expanded: {exp_ucs}")
print(f"Path: {path_ucs}")

--- 9a. Search Results for Goal (9, 9) ---

[A* Search]
Path Length: 18 steps
Nodes Expanded: 72
Path: [(0, 0), (1, 0), (2, 0), (2, 1), (3, 1), (3, 2), (3, 3), (3, 4), (3, 5), (3, 6), (3, 7), (3, 8), (4, 8), (5, 8), (5, 9), (6, 9), (7, 9), (8, 9), (9, 9)]

[Greedy Best First Search (GBF)]
Path Length: 18 steps
Nodes Expanded: 19
Path: [(0, 0), (1, 0), (2, 0), (3, 0), (4, 0), (5, 0), (6, 0), (7, 0), (8, 0), (9, 0), (9, 1), (9, 2), (9, 3), (9, 4), (9, 5), (9, 6), (9, 7), (9, 8), (9, 9)]

[Uniform Cost Search (UCS)]
Path Length: 18 steps
Nodes Expanded: 162
Path: [(0, 0), (1, 0), (2, 0), (3, 0), (4, 0), (5, 0), (6, 0), (6, 1), (7, 1), (7, 2), (8, 2), (8, 3), (8, 4), (9, 4), (9, 5), (9, 6), (9, 7), (9, 8), (9, 9)]


# 9b. Compare A* with Greedy Best First Search and Uniform Cost Search. What is the advantage of A*?

A* is basically the perfect middle ground between the other two. Its main advantage over Uniform Cost Search (UCS) is efficiency: A* uses the heuristic to "aim" towards the goal, so it expanded way fewer nodes (72 vs. 162) to find the same length path.

Its advantage over Greedy Best First (GBF) is safety/optimality. Even though GBF found a path of 18 steps in this specific maze, it isn't guaranteed to find the shortest path in every scenario because it blindly rushes the goal. A* is guaranteed to be optimal because it balances the cost traveled so far with the estimated distance remaining.

# 9c. What is the advantage of Greedy Best First Search?

The big advantage of Greedy Best First Search is speed and low memory usage. In my results, GBF was by far the fastest, expanding only 19 nodes compared to 72 for A* and 162 for UCS. Because it ignores the cost of the path taken so far and just greedily chooses the neighbor closest to the goal, it can "zoom" straight there if the path is relatively clear, making it very computationally efficient.